### DISCUSSION TITLE
[BUG] Pay::Customer#customer no longer exists

### DISCUSSION THREAD

**Discussion by schristm** [pay-rails/pay#1109] [Q&A] [answered]
Looks like the `customer` method used to access the payment processor's customer data is no longer available after this change:

https://github.com/pay-rails/pay/commit/1c1a5c1ffa12d1f6579a2eff6b89d5a48924221b#diff-cd9fb0877977dacc88d88dac941f3cf5bd86d647c76c70c8138331e559f852fcL24

Documentation suggests it should still work:

https://github.com/pay-rails/pay/blob/main/docs/3_customers.md#retrieving-a-customer-object-from-the-payment-processor

Also tried `api_record` to no avail.

Thanks!

**Comment by excid3** [comment]
Correct, it is now `user.payment_processor.api_record`. I've updated the docs. 👍

**Comment by schristm** [comment]
Hey Chris,

Thanks for getting back to me. I'm still having some difficulty with this change. I'm specifically working on a migration that moves subscriptions from a legacy system to `pay`.

This does not work:
```ruby
Pay::Customer.first.api_record # => undefined method `api_record' for #<Pay::Customer id: 1, owner_type: "Account", owner_id: 1, processor: "stripe", processor_id: "redacted", default: true, data: {"currency"=>"usd"}, stripe_account: nil, deleted_at: nil, created_at: "2020-10-13 06:55:17.768998000 +0000", updated_at: "2024-10-21 07:13:38.061001000 +0000"> (NoMethodError)
```

This does retrieve the Stripe data:
```ruby
Pay::Stripe::Customer.first.api_record
```

However, this fails:
```ruby
card_id = Pay::Stripe::Customer.first.api_record.default_source
Pay::Stripe::PaymentMethod.sync(card_id) # => undefined method `api_record' for #<Pay::Customer id: 1, owner_type: "Account", owner_id: 1, processor: "stripe", processor_id: "redacted", default: true, data: {"currency"=>"usd"}, stripe_account: nil, deleted_at: nil, created_at: "2020-10-13 06:55:17.768998000 +0000", updated_at: "2024-10-21 07:13:38.061001000 +0000"> (NoMethodError)
```

I think it's because the `Pay::Stripe::PaymentMethod.sync` method is still using `Pay::Customer` internally instead of `Pay::Stripe::Customer`, and `Pay::Customer` doesn't seem to support `api_record`.

Which tells me this probably doesn't work either (haven't tested):

https://github.com/pay-rails/pay/blob/3a5580c1d931fbbe73aed36107ec09d96255882a/lib/tasks/pay.rake#L22

**Reply by excid3** [reply, marked as answer]
`Pay::Customer` is an abstract class for single table inheritance, so you will need a `Pay::PROCESSOR::Customer` object to ask for API records.

You will get a concrete class back when you use `Pay::Customer.first`:
```irb
dummy(test)> Pay::Customer.first
=> #<Pay::Stripe::Customer:0x0000000129b5c7d0>
```

If you don't, your database needs to be fixed and the `type` column set on each record. This can happen if you upgraded from an older Pay but didn't backfill the columns as per [UPGRADE.md](https://github.com/pay-rails/pay/blob/main/UPGRADE.md) which installs the new migration in your app: https://github.com/pay-rails/pay/blob/main/db/migrate/2_add_pay_sti_columns.rb

**Reply by schristm** [reply]
Thanks, @excid3. I did miss that step.
